naming/README.md

```diff
@@ -0,0 +1,93 @@
+
+- Domain name declarations need to be verified by the owner associated to the parent domain. That owner can also be set to Null, removing the permissions needed for a specific domain name and its sub-domains. The declaration of a new TLD (top-level domain) name needs to be verified by the root TLD owner as it is a sub-domain of it. Therefore the owner of the root domain can administrate the TLD and sub-domain distribution.
+Similarly to DNS, we use `"."` as a root domain name seed. Nevertheless, as we need to protect the domain names from collisions with DNS names, the root tld is written out as the string `".sol"` for url lookups in the browser search bar via the extension.
+For example the name resolution of `something.sol` would use the seeds: `[["something"], ["sol"], ["."]]` (converting the strings to bytes of course) but would be needed to be written as `something.sol.sol` in the browser bar.
+
+- Twitter handles can be added as sub-domain names of one TLD name the owner of which will therefore hold the right to add a twitter handle name. This enables the verification of twitter accounts for example by asking the user to tweet his pubkey or a signed message. A bot that holds the private twitter-domain-name owner key can then sign the AddName instruction and send it back to the user who will then submit it to the program. This verification bot acts as a kind of centralized oracle that verifies the names. For now a centralized set up seems appropriate for this. Additional js bindings that structure this will be implemented.
+
+- [Proposal] An additional instruction (or another contract) could do the following: mint an NFT for an already existing name and activate ownership for the holder of the NFT only. This would enable easy trading of names.
+
+
+Program Interface:
+
+```
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = u16;
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct NameState {
+    status: u8,
+    values: Vec<(ValueType, String)>,
+    owner_pubkey: Option<Pubkey>,
+    owner_name: Option<String>,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum NameInstruction {
+    /// Add a record for the given name to the naming service.
+    /// The owner can be specified as a pubkey or (exclusively) as another name (the owner then being
+    /// the owner of that name, this can be usefull for domain name extensions).
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   1. `[writable]` The program account derived from the name
+    ///   2. `[signer]` (Optional) The domain name parent account if the given name is a sub-domain
+    ///   3...N `[]` The list of accounts needed to find an owner which is a pubkey or a null
+    AddName {
+        name: String,
+        values: Vec<(ValueType, String)>,
+        owner_pubkey: Option<Pubkey>,
+        owner_name: Option<String>,
```

Comment:
If these are exclusive a single enum fields would make more sense
```rust
enum OwnerKind {
    Account(Pubkey),
    Name(String),
    None,
}
```

Though I'm not sure I see the value in the name variant :thinking:

Reply:
The name variant is useful when changing an authority on one record with implicit propagation to other records which mention its name as their authority. Imagine having authority over a certain domain and seamlessly transferring it and all of its subdomains in one transaction. This allows for larger trustless transactions.

Reply:
> The name variant is useful when changing an authority on one record with implicit propagation to other records which mention its name as their authority. Imagine having authority over a certain domain and seamlessly transferring it and all of its subdomains in one transaction. This allows for larger trustless transactions.

It seems like this can be done just as easily with by querying the program accounts by authority from the client side (similar to `getTokenAccountsByOwner` for SPL Token